[RayService] Stable Diffusion example (#1181)
ray-project/kuberay: docs/guidance/aws-eks-gpu-cluster.md

> # Prepare Amazon EKS cluster with GPU for KubeRay examples

Reviewer: Could you retitle this to make it more generic? How about `Start Amazon EKS Cluster with GPUs for KubeRay`?

Author: https://github.com/ray-project/kuberay/pull/1181/commits/4d018b2fb5561bffb45c83eb0bcf6b7c9d1db9f5

## Step 1: Create a Kubernetes cluster on Amazon EKS

Follow the first two steps in this [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#)

> to: (1) Create your Amazon EKS cluster (2) Configure your computer to communicate with your cluster.

Reviewer: ```suggestion
to: (1) create your Amazon EKS cluster and (2) configure your local machine to communicate with your cluster.
```

Author: These are the names of the sections in the following URL: https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#.

<img width="1209" alt="Screen Shot 2023-06-22 at 10 57 02 AM" src="https://github.com/ray-project/kuberay/assets/20109646/85d0d4fb-75e8-49ab-83f3-bba9a54761fa">

Reviewer: Oh got it, let's keep it as-is then.

## Step 2: Create node groups for the Amazon EKS cluster

> You can follow "Step 3: Create nodes" in this [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#) to create node groups. The following section provides more detailed information.

Reviewer: ```suggestion
You can follow "Step 3: Create nodes" in [this AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#) to create node groups. The following section provides more detailed information.
```

### Create a CPU node group

> Typically, we do not run GPU workloads on the Ray head. Create a CPU node group for all Pods except Ray GPU 

Reviewer: Could you give some more details on how large this node group should be? And could you specify what size node users should be using. You could do this by attaching a screenshot showing a working configuration from the AWS console.

Author: This document will be used for multiple examples, each may require a different setup. Take this PR as an example, I have provided additional details in Step 1 of [stable-diffusion-rayservice.md](https://github.com/ray-project/kuberay/pull/1181/files#diff-e2830c575dbdd61d5c4f45500ef2a56f0ef4777acccfb695a508d20dcd5224ca). 

(1) Do not provide details in [aws-eks-gpu-cluster.md](https://github.com/ray-project/kuberay/pull/1181/files/0eb5db1e2353f8ad9e38ef3a9b595cb350bf6ff5#diff-db316438b08e0a251e5c17308a77017d75ec589bdfaa716fef765aff981d3f8d). Instead, include the details in the example documentation. (current method)
(2) Provide a common configuration in this document, and if the example does not fit the configuration, include additional details in the example documentation.

Do you have any preference?

Reviewer: I'd prefer (2). I don't want users start Ray with nodes that are too small and having it never start.

Author: Updated

workers, such as the KubeRay operator, Ray head, and CoreDNS Pods.

### Create a GPU node group

> Create a GPU node group for Ray GPU workers. Add a Kubernetes taint to prevent scheduling CPU Pods on this GPU node group. For KubeRay examples, we add the following taint to the GPU nodes: `Key: ray.io/node-type, Value: worker, Effect: NoSchedule`.

Reviewer: Could you add what size GPU nodes users should be using?

> Create a GPU node group for Ray GPU workers. Add a Kubernetes taint to prevent scheduling CPU Pods on this GPU node group. For KubeRay examples, we add the following taint to the GPU nodes: `Key: ray.io/node-type, Value: worker, Effect: NoSchedule`.

Author: This taint is only applicable to KubeRay examples that set corresponding `tolerations` settings for GPU Pods. If users do not specify tolerations, no Pod will be able to be scheduled on these nodes. Adding the taint is not enough. Hence, we may need to mention KubeRay examples here. HDYT? We may also need to mention `tolerations` here.

> Create a GPU node group for Ray GPU workers. Add a Kubernetes taint to prevent scheduling CPU Pods on this GPU node group. For KubeRay examples, we add the following taint to the GPU nodes: `Key: ray.io/node-type, Value: worker, Effect: NoSchedule`.

Author: Updated

Next, **please follow Step 4 to install the NVIDIA device plugin.**

> Warning: GPU nodes are extremely expensive. Please remember to delete the cluster if you no longer need it.

## Step 3: Verify the node groups

> If you encounter permission issues with `eksctl`, you can navigate to your AWS account's webpage and copy the

Reviewer: Could you make this a note?

Author: https://github.com/ray-project/kuberay/pull/1181/commits/4d018b2fb5561bffb45c83eb0bcf6b7c9d1db9f5

credential environment variables, including `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN`,
from the "Command line or programmatic access" page.

```sh
eksctl get nodegroup --cluster ${YOUR_EKS_NAME}

# CLUSTER         NODEGROUP       STATUS  CREATED                 MIN SIZE        MAX SIZE        DESIRED CAPACITY        INSTANCE TYPE   IMAGE ID                        ASG NAME                           TYPE
# ${YOUR_EKS_NAME}     cpu-node-group  ACTIVE  2023-06-05T21:31:49Z    0               1               1                       m5.xlarge       AL2_x86_64                      eks-cpu-node-group-...     managed
# ${YOUR_EKS_NAME}     gpu-node-group  ACTIVE  2023-06-05T22:01:44Z    0               1               1                       g5.12xlarge     BOTTLEROCKET_x86_64_NVIDIA      eks-gpu-node-group-...     managed
```

> ## Step 4: Install the DaemonSet for NVIDIA device plugin for Kubernetes

Reviewer: Can you briefly explain why users should do this?

Author: Copy a sentence from https://github.com/NVIDIA/k8s-device-plugin.

https://github.com/ray-project/kuberay/pull/1181/commits/fa89e625fe9f3c606b319cd8c76c0b734be28158

> If you encounter permission issues with `kubectl`, you can follow "Step 2: Configure your computer to communicate with your cluster"

Reviewer: Could you make this a note?

Author: https://github.com/ray-project/kuberay/pull/1181/commits/4d018b2fb5561bffb45c83eb0bcf6b7c9d1db9f5

# Start Amazon EKS Cluster with GPUs for KubeRay

## Step 1: Create a Kubernetes cluster on Amazon EKS

Follow the first two steps in [this AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#)

> to: (1) Create your Amazon EKS cluster (2) Configure your computer to communicate with your cluster.

Reviewer: [Nit] Adding `and` here makes the sentence read smoother.
```suggestion
to: (1) create your Amazon EKS cluster and (2) configure your computer to communicate with your cluster.
```

## Step 2: Create node groups for the Amazon EKS cluster

> You can follow "Step 3: Create nodes" in [this AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#) to create node groups. The following section provides more detailed information.

Reviewer: ```suggestion
Follow "Step 3: Create nodes" in [this AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#) to create node groups. The following section provides more detailed information.
```

> You can follow "Step 3: Create nodes" in [this AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#) to create node groups. The following section provides more detailed information.

Reviewer: Generally, should Pods be capitalized or lowercase?

Author: Kubernetes's website uses "Pod".

### Create a CPU node group

> Typically, we do not run GPU workloads on the Ray head. Create a CPU node group for all Pods except Ray GPU 

Reviewer: ```suggestion
Typically, avoid running GPU workloads on the Ray head. Create a CPU node group for all Pods except Ray GPU 
```

workers, such as the KubeRay operator, Ray head, and CoreDNS Pods.

> * A common configuration working for most KubeRay examples:

Reviewer: ```suggestion
Here's a common configuration that works for most KubeRay examples in the docs:
```

  * Instance type: [**m5.xlarge**](https://aws.amazon.com/ec2/instance-types/m5/) (4 vCPU; 16 GB RAM)
  * Disk size: 256 GB
  * Desired size: 1, Min size: 0, Max size: 1

### Create a GPU node group

Create a GPU node group for Ray GPU workers.

> 1. A common configuration working for most KubeRay examples:

Reviewer: ```suggestion
1. Here's a common configuration that works for most KubeRay examples in the docs:
```

   * AMI type: Bottlerocket NVIDIA (BOTTLEROCKET_x86_64_NVIDIA)
   * Instance type: [**g5.xlarge**](https://aws.amazon.com/ec2/instance-types/g5/) (1 GPU; 24 GB GPU Memory; 4 vCPUs; 16 GB RAM)
   * Disk size: 1024 GB
   * Desired size: 1, Min size: 0, Max size: 1

> 2. **Please follow Step 4 to install the NVIDIA device plugin.**

Reviewer: ```suggestion
2. If you don't use the `Bottlerocket NVIDIA` AMI type, you may need to install the NVIDIA device plugin.
```

>    * If you use `AMI type: Bottlerocket NVIDIA`, there is no need to install NVIDIA device plugin.

Reviewer: ```suggestion
   * Please follow Step 4 to install the NVIDIA device plugin.
```

>    * For other AMI types, you may need to install the NVIDIA device plugin DaemonSet in order to run GPU-enabled containers in your Amazon EKS cluster.

Reviewer: ```suggestion
   * The NVIDIA device plugin DaemonSet helps run GPU-enabled containers in your Amazon EKS cluster.
```

>    If the GPU nodes have taints, you need to add `tolerations` to `nvidia-device-plugin.yml` to enable the DaemonSet to schedule Pods on the GPU nodes."

Reviewer: ```suggestion
   If the GPU nodes have taints, add `tolerations` to `nvidia-device-plugin.yml` to enable the DaemonSet to schedule Pods on the GPU nodes."
```

> 3. Add a Kubernetes taint to prevent scheduling CPU Pods on this GPU node group. For KubeRay examples, we add the following taint to the GPU nodes: `Key: ray.io/node-type, Value: worker, Effect: NoSchedule`, and include the corresponding `tolerations` for GPU Ray worker Pods.

Reviewer: ```suggestion
3. Add a Kubernetes taint to prevent scheduling CPU Pods on this GPU node group. For KubeRay examples, add the following taint to the GPU nodes: `Key: ray.io/node-type, Value: worker, Effect: NoSchedule`, and include the corresponding `tolerations` for GPU Ray worker Pods.
```

> Warning: GPU nodes are extremely expensive. Please remember to delete the cluster if you no longer need it.

## Step 3: Verify the node groups

> > **Note:** If you encounter permission issues with `eksctl`, you can navigate to your AWS account's webpage and copy the

Reviewer: ```suggestion
> **Note:** If you encounter permission issues with `eksctl`, navigate to your AWS account's webpage and copy the
```

credential environment variables, including `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN`,
from the "Command line or programmatic access" page.

```sh
eksctl get nodegroup --cluster ${YOUR_EKS_NAME}

# CLUSTER         NODEGROUP       STATUS  CREATED                 MIN SIZE        MAX SIZE        DESIRED CAPACITY        INSTANCE TYPE   IMAGE ID                        ASG NAME                           TYPE
# ${YOUR_EKS_NAME}     cpu-node-group  ACTIVE  2023-06-05T21:31:49Z    0               1               1                       m5.xlarge       AL2_x86_64                      eks-cpu-node-group-...     managed
# ${YOUR_EKS_NAME}     gpu-node-group  ACTIVE  2023-06-05T22:01:44Z    0               1               1                       g5.12xlarge     BOTTLEROCKET_x86_64_NVIDIA      eks-gpu-node-group-...     managed
```

## Step 4: Install the DaemonSet for NVIDIA device plugin

> 
> > **Note:** If you encounter permission issues with `kubectl`, you can follow "Step 2: Configure your computer to communicate with your cluster"

Reviewer: ```suggestion

> **Note:** If you encounter permission issues with `kubectl`, follow "Step 2: Configure your computer to communicate with your cluster"
```
